### ISSUE TITLE
Epic - Can't claim 18+ game (reaching timeout)

### ISSUE THREAD

**Issue by maxp1110** [vogler/free-games-claimer#120]
Hello,
I am currently running into a problem with this weeks free giveaway at Epic (Dying Light Enhanced Edition). The script seems to be not able to claim the game and runs into an timeout.

```
Free games: [
  'https://store.epicgames.com/en-US/p/dying-light',
  'https://store.epicgames.com/en-US/p/shapez-0929c9'
]
  This game contains mature content recommended only for ages 18+
Current free game: Dying Light Enhanced Edition
  Not in library yet! Click GET.
locator.click: Timeout 60000ms exceeded.
=========================== logs ===========================
waiting for frameLocator('#webPurchaseContainer iframe').locator('button:has-text("Place Order"):not(:has(.payment-loading--loading))')
============================================================
    at file:///fgc/epic-games.js:180:99 {
  name: 'TimeoutError'
}
```

Is there a way for me to fix this issue?

Thanks!

**Comment by OJ7** [comment]
Receiving the same error with slightly different logs:

```
Free games: [
  'https://store.epicgames.com/en-US/p/dying-light',
  'https://store.epicgames.com/en-US/p/shapez-0929c9'
]
  This game contains mature content recommended only for ages 18+
page.click: Timeout 60000ms exceeded.
=========================== logs ===========================
waiting for locator('button:has-text("Continue")')
  locator resolved to <button class="css-1a6we1t">…</button>
attempting click action
  waiting for element to be visible, enabled and stable
  element is visible, enabled and stable
  scrolling into view if needed
  done scrolling
  performing click action
============================================================
    at file:///fgc/epic-games.js:120:18 {
  name: 'TimeoutError'
}
```

In my case, when clicking the "GET" button, the popup shows a gray screen with a spinner, but eventually dismisses and the "GET" button is clickable again. I was not able to redeem it using VNC by manually clicking either (same issue). 

I haven't tried manually redeeming it in my regular browser yet in case a new version needs to be tested for this.

**Comment by xBlackPoison357x** [comment]
![image](https://user-images.githubusercontent.com/4401213/230505380-b2aa018e-fc7c-4d51-b07a-1c20638e90a1.png)
Same error

**Comment by EthanC** [comment]
I have this error too. Interestingly enough, it eventually succeeded in claiming the title without any modifications.

**Logs**

```
Free games: [
  'https://store.epicgames.com/en-US/p/dying-light',
  'https://store.epicgames.com/en-US/p/shapez-0929c9'
]
  This game contains mature content recommended only for ages 18+
page.click: Timeout 60000ms exceeded.
=========================== logs ===========================
waiting for locator('button:has-text("Continue")')
  locator resolved to <button class="css-1a6we1t">…</button>
attempting click action
  waiting for element to be visible, enabled and stable
  element is visible, enabled and stable
  scrolling into view if needed
  done scrolling
  performing click action
============================================================
    at file:///fgc/epic-games.js:120:18 {
  name: 'TimeoutError'
}
```

**Comment by vogler** [comment]
Can't test it, since the game is not available for german accounts:
![image](https://user-images.githubusercontent.com/493741/230563630-71156004-1431-4c8f-9a32-7602c5d6bc08.png)

There's already code that should deal with this message:
https://github.com/vogler/free-games-claimer/blob/6cbd9fe5e3245b511d4424df82ea9b143b7da4cf/epic-games.js#L117-L121

Observations:
1. Since the code succeeded unmodified for @EthanC after some time, it must be some timing issue.
2. All of your logs show the `This game contains mature content ...` message, so there's no timing issue detecting that there's a `Continue` button to click.
3. For @maxp1110, it did successfully click `Continue` but timed out later; for @OJ7, @xBlackPoison357x, @EthanC it timed out clicking this `Continue` button.
4. All of Playwright's preconditions for click seem to be met: `element is visible, enabled and stable`. According to the [docs](https://playwright.dev/docs/api/class-locator#locator-click), it could only be the last step `Wait for initiated navigations to either succeed or fail, unless noWaitAfter option is set.`

There's a couple of things we can try:
1. Maybe it's the same as for some past issue where the correct event handler has not yet been attached, which was fixed by adding a click delay.
2. Maybe there's some issue with loading a resource after, which could be fixed by `noWaitAfter`.
3. If all else fails, we could add some delays around actions and see if that helps.

**Comment by EthanC** [comment]
Additional context - the script continued to sometimes TimeoutError after successfully claiming the game.

![image](https://user-images.githubusercontent.com/16727756/230569587-99ef4370-114a-412d-86e2-57e684539ca7.png)

I've now updated to the latest Docker image following the above commit. 4/10 attempts I received a new error...

```
Free games: [
  'https://store.epicgames.com/en-US/p/dying-light',
  'https://store.epicgames.com/en-US/p/shapez-0929c9'
]
  This game contains mature content recommended only for ages 18+
Current free game: Dying Light Enhanced Edition
  Already in library! Nothing to claim.
Current free game: shapez
  Already in library! Nothing to claim.
node:internal/process/promises:289
            triggerUncaughtException(err, true /* fromPromise */);
            ^
ProtocolError: Target closed
    at /fgc/node_modules/playwright-core/lib/server/firefox/ffConnection.js:189:16
    at new Promise (<anonymous>)
    at FFSession.send (/fgc/node_modules/playwright-core/lib/server/firefox/ffConnection.js:185:12)
    at RawMouseImpl.move (/fgc/node_modules/playwright-core/lib/server/firefox/ffInput.js:91:24)
    at Mouse.move (/fgc/node_modules/playwright-core/lib/server/input.js:153:23)
    at Mouse.click (/fgc/node_modules/playwright-core/lib/server/input.js:180:12)
    at /fgc/node_modules/playwright-core/lib/server/dom.js:468:109
    at /fgc/node_modules/playwright-core/lib/server/dom.js:430:13
    at FrameManager.waitForSignalsCreatedBy (/fgc/node_modules/playwright-core/lib/server/frames.js:114:26)
    at ElementHandle._performPointerAction (/fgc/node_modules/playwright-core/lib/server/dom.js:424:57) {
  sessionClosed: true
}
Node.js v19.8.1
```

All other attempts were completed successfully. I'll report back if I see the TimeoutError again.

**Comment by vogler** [comment]
@EthanC Ok, so the click delay seems to have solved the Timeout for clicking on Continue. The `Target closed` error means that the browser closed while there were still promises to be resolved. In case of the already claimed game, this should be the click on Continue, which is strange since it has an `await` which should block. Maybe some internal Playwright issue.
I added a delay after the click as well, which hopefully is enough to avoid this issue.

**Comment by maxp1110** [comment]
Hi @vogler: Thanks for the updated version! After the 1st attempt still showed the same error message, the 2nd try gave probably the reason why it did not work:
```
Current free game: Dying Light Enhanced Edition
  Not in library yet! Click GET.
  This product is unavailable in your region!
```
Im also from Germany, however my epic account is set to US and there seems to be IP-Geoblock when trying to purchase it, even when trying to do so manually 
![image](https://user-images.githubusercontent.com/130100098/230581925-caf277e6-d585-4173-881c-b2429dfb2dca.png)

**Comment by OJ7** [comment]
I'm reading on a forum that others had issues yesterday redeeming it too while manually trying to claim the game. I just ran the script again now and it worked the first time. Looks like this may have been a temporary issue with Epic's website.

**Comment by vogler** [comment]
> edit: I know this feature might probably not be of interest for the most folks, but could VPN-support be added to the docker?

I opened an issue for it.

**Comment by xBlackPoison357x** [comment]
> @EthanC Ok, so the click delay seems to have solved the Timeout for clicking on Continue. The `Target closed` error means that the browser closed while there were still promises to be resolved. In case of the already claimed game, this should be the click on Continue, which is strange since it has an `await` which should block. Maybe some internal Playwright issue. I added a delay after the click as well, which hopefully is enough to avoid this issue.

Just got the target closed error today.
![image](https://user-images.githubusercontent.com/4401213/230992114-dcf949c8-408a-4e20-821f-7cb8aa641e24.png)

**Comment by vogler** [comment]
@xBlackPoison357x did you also get a notification for it?

It's a bit annoying that the trace doesn't contain any useful information, but it should be the click on Continue which somehow hangs around after the promise already resolved.

I assume claiming 18+ games should now work for everyone without a timeout. Just this `Target closed` exception left which could also be ignored.

**Comment by vogler** [comment]
Just tested Mordhau (18+):
1. Game was claimed successfully, but got Timeout `waiting for locator('text=Thank you for buying') to be visible`
2. Got `Target closed` exception sometimes when running afterwards.

Regarding 2., the problem is between clicking on the `Continue` button:
https://github.com/vogler/free-games-claimer/blob/351bf00c7b78a94b72d2a5fb1d9085b385ec41bb/epic-games.js#L120
and closing the browser context at the end:
https://github.com/vogler/free-games-claimer/blob/351bf00c7b78a94b72d2a5fb1d9085b385ec41bb/epic-games.js#L229
When commenting either out, I did not get `Target closed`.
However, without the click, the message stays there and the title of the game is not detected correctly, and without the close, it just waits around at the end.
Same effect when I removed the `, { delay: 111 }` from the click. Adding `noWaitAfter` had no effect.
So it seems like Playwright creates another promise with `delay` which then hangs around waiting.
Catching the exception to ignore it did not work - neither adding `.catch(_ => { })` for the click, nor
```js
try {
  await context.close();
} catch (error) {
  console.error(error.message, error);
}
```

**Comment by ChekeredList71** [comment]
For me it fails if an 18+ game is available, but I'm not sure about this, because the first game it tries to claim isn't 18+.
Could it be, that it is the same issue, or should it go into a different issue? Thanks.

```
2023-04-29 11:48:32.500 started checking epic-games
Signed in as ChekeredList71
Free games: [
  'https://store.epicgames.com/en-US/p/breathedge',
  'https://store.epicgames.com/en-US/p/poker-club-6e18ee'
]
Current free game: Breathedge
  Not in library yet! Click GET.
locator.click: Timeout 60000ms exceeded.
=========================== logs ===========================
waiting for frameLocator('#webPurchaseContainer iframe').locator('button:has-text("Place Order"):not(:has(.payment-loading--loading))')
============================================================
    at file:///fgc/epic-games.js:184:99 {
  name: 'TimeoutError'
}
node:internal/process/promises:289
            triggerUncaughtException(err, true /* fromPromise */);
            ^

ProtocolError: Target closed
    at /fgc/node_modules/playwright-core/lib/server/firefox/ffConnection.js:189:16
    at new Promise (<anonymous>)
    at FFSession.send (/fgc/node_modules/playwright-core/lib/server/firefox/ffConnection.js:185:12)
    at RawMouseImpl.move (/fgc/node_modules/playwright-core/lib/server/firefox/ffInput.js:91:24)
    at Mouse.move (/fgc/node_modules/playwright-core/lib/server/input.js:153:23)
    at Mouse.click (/fgc/node_modules/playwright-core/lib/server/input.js:180:12)
    at /fgc/node_modules/playwright-core/lib/server/dom.js:468:109
    at /fgc/node_modules/playwright-core/lib/server/dom.js:430:13
    at FrameManager.waitForSignalsCreatedBy (/fgc/node_modules/playwright-core/lib/server/frames.js:114:26)
    at ElementHandle._performPointerAction (/fgc/node_modules/playwright-core/lib/server/dom.js:424:57) {
  sessionClosed: true
}

Node.js v19.9.0
``` 

OS: Linux Mint 21.1

**Comment by vogler** [comment]
@ChekeredList71 No, this is something else, neither are 18+ (at least for me in DE).
Check via VNC what's up and open a new issue with the details if the problem persists.

**Comment by xBlackPoison357x** [comment]
![image](https://github.com/vogler/free-games-claimer/assets/4401213/09a4b0dc-3dfd-4daa-bddd-b42904b7d302)


Just got a Target closed.

**Comment by vogler** [comment]
Closing this since timeout is fixed. Tracking `Target closed` exception in #170.
